tests/integration/test_jobmanager.py

```diff
@@ -166,14 +168,140 @@ async def test_jobmanager(
+        **base_builder_agent_health_request | json_executing
+    ).respond_with_data("OK")
+    httpserver.expect_oneshot_request(
+        **base_builder_agent_health_request | json_finished
+    ).respond_with_data("OK")
+    httpserver.expect_request(
+        **base_builder_agent_health_request | json_finished
+    ).respond_with_data("OK")
 
-    # At this point the openstack instance is spawned, but cloud init is not yet correct.
-    # The reconcile loop is still not adapted and will kill the instance incorrectly.
-    assert True, "At this point the builder should be spawned, but pending to replace cloud init."
+    with httpserver.wait(raise_assertions=False, stop_on_nohandler=False, timeout=120) as waiting:
+        logger.info("Waiting for builder-agent to contact us.")
+    logger.info("server log after executing: %s ", (httpserver.log))
+    assert waiting.result, "builder-agent did not execute or finished."
+
+    httpserver.check_assertions()
+    # The reconcile loop is still not adapted and will badly kill the instance as the ssh
+    # health check will mark the instance as unhealthy.
+
+
+async def _prepare_runner_tunnel_for_builder_agent(
+    instance_helper, unit, jobmanager_address, jobmanager_port
+) -> bool:
+    """Prepare the runner tunner so the builder-agent can access the fake jobmanager.
+
+    This function return False if the tunnel could not be prepared and
+    retrying is possible.
+    """
+    logger.info("trying to prepare tunnel for builder agent")
+    try:
+        server = instance_helper.get_single_runner(unit)
+    except AssertionError:
+        # No runner or two runners
+        logger.info("no runner or two runners in unit, return False")
+        return False
+    network_address_list = server.addresses.values()
+    if not network_address_list:
+        logger.info("no addresses yet, return False")
+        return False
+
+    exit_code, stdout, _ = await instance_helper.run_in_instance(
+        unit,
+        "'echo hello'",
+        timeout=10,
+    )
+    if exit_code != 0 or not stdout or "hello" not in stdout:
+        logger.info("cannot ssh yet, return False")
+        return False
+
+    # Not sure about this. We should check if the nftables interfere with the iptables rules.
+    # For now, a sleep for a bit of time so we the runner has time to flush and apply the
+    # nftables rules. We may also check this issue in other tests.
+    await asyncio.sleep(60)
+
+    dnat_comman_in_runner = f"sudo iptables -t nat -A OUTPUT -p tcp -d {jobmanager_address} --dport {jobmanager_port} -j DNAT --to-destination 127.0.0.1:{jobmanager_port}"  # noqa  # pylint: disable=line-too-long
+    exit_code, _, _ = await instance_helper.run_in_instance(
+        unit,
+        dnat_comman_in_runner,
+        timeout=10,
+    )
+    assert exit_code == 0, "could not apply iptables"
+    await instance_helper.expose_to_instance(
+        unit=unit, port=jobmanager_port, host=jobmanager_address
+    )
```

Comment:
I've understood it but can you add an explanatory comment (that we redirect traffic to a local port which is relayed using ssh remote port forwarding implemented by expose_to_instance), I think this might be hard to read for external people.